onnxruntime/core/providers/cpu/ml/tree_ensemble_aggregator.h

```diff
@@ -0,0 +1,481 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include "core/common/common.h"
+#include "core/framework/op_kernel.h"
+#include "ml_common.h"
+#include <math.h>
+
+namespace onnxruntime {
+namespace ml {
+
+struct TreeNodeElementId {
+  int tree_id;
+  int node_id;
+  inline bool operator==(const TreeNodeElementId& xyz) const {
+    return (tree_id == xyz.tree_id) && (node_id == xyz.node_id);
+  }
+  inline bool operator<(const TreeNodeElementId& xyz) const {
+    return ((tree_id < xyz.tree_id) || (tree_id == xyz.tree_id && node_id < xyz.node_id));
+  }
+};
+
+template <typename T>
+struct SparseValue {
+  int64_t i;
+  T value;
+};
+
+enum MissingTrack {
+  NONE,
+  TRUE,
+  FALSE
+};
+
+template <typename T>
+struct TreeNodeElement {
+  TreeNodeElementId id;
+  int feature_id;
+  T value;
+  T hitrates;
+  NODE_MODE mode;
+  TreeNodeElement<T>* truenode;
+  TreeNodeElement<T>* falsenode;
+  MissingTrack missing_tracks;
+  std::vector<SparseValue<T>> weights;
+
+  bool is_not_leave;
+  bool is_missing_track_true;
+};
+
+template <typename ITYPE, typename OTYPE>
+class _Aggregator {
+ protected:
```

Comment:
>protected [](start = 1, length = 9)

private

Reply:
Some members are accessed by children classes.

Reply:
You can have protected accessors for children to use such things. Direct access means the parent class can't guarantee any invariants it may want to apply to these members. Also means there are a lot more places that can change the value of the member vs. that only happening in the code for the parent class (assuming changes by the children are done using a protected method in the parent class). 

This can be cleaned it up later. Let's get the changes in and see where we're at with the benchmarks.